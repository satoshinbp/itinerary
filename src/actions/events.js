import moment from 'moment'

export const addEvent = (
  {
    id='',
    tripId = '',
    title = '',
    date = 0,
    startTime = 0,
    endTime = 0,
    location = '',
    note = ''
  } = {}
) => ({
  type: 'ADD_EVENT',
  event: {
    id,
    tripId,
    title,
    date,
    startTime,
    endTime,
    location,
    note
  }
})

export const editEvent = (id, updates) => ({
  type: 'EDIT_EVENT',
  id,
  updates
})

export const removeEvent = (id = '') => ({
  type: 'REMOVE_EVENT',
  id
})