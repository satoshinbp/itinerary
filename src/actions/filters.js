export const showUpcomingTrip = () => ({
  type: 'SHOW_UPCOMING_TRIP',
})

export const showPastTrip = () => ({
  type: 'SHOW_PAST_TRIP',
})

export const showAllTrip = () => ({
  type: 'SHOW_ALL_TRIP',
})