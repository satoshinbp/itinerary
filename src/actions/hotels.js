export const addHotel = (
  {
    id='',
    tripId = '',
    name = '',
    checkInDate = 0,
    checkOutDate = 0,
    checkInTime = 0,
    checkOutTime = 0,
    ETA = 0,
    ETD = 0,
    location ='',
    note = ''
  } = {}
) => ({
  type: 'ADD_HOTEL',
  hotel: {
    id,
    tripId,
    name,
    checkInDate,
    checkOutDate,
    checkInTime,
    checkOutTime,
    ETA,
    ETD,
    location,
    note
  }
})

export const editHotel = (id, updates) => ({
  type: 'EDIT_HOTEL',
  id,
  updates
})

export const removeHotel = (id = '') => ({
  type: 'REMOVE_HOTEL',
  id
})