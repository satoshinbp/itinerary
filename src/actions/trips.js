import moment from 'moment'

export const addTrip = (
  {
    id = '',
    title = '',
    startDate = 0,
    endDate = 0,
    note = ''
  } = {}
) => ({
  type: 'ADD_TRIP',
  trip: {
    id,
    title,
    startDate,
    endDate,
    note
  }
})

export const editTrip = (id, updates) => ({
  type: 'EDIT_TRIP',
  id,
  updates
})

export const removeTrip = (id = '') => ({
  type: 'REMOVE_TRIP',
  id
})