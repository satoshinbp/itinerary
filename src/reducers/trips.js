const tripsReducerDefaultState = [];

export default (state = tripsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_TRIP':
      return [
        ...state,
        action.trip
      ];
    case 'EDIT_TRIP':
      return state.map((trip) => {
        if (trip.id === action.id) {
           return {
            ...trip,
            ...action.updates
           }
        } else {
          return trip;
        }
      });
    case 'REMOVE_TRIP':
      return state.filter(({ id }) => id !== action.id);
    default:
      return state;
  }
};