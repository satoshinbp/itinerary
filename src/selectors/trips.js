import moment from 'moment';

export default (trips, { filter }) => {
  return trips.filter((trip) => {
    switch (filter) {
      case 'upcoming':
        return moment(trip.endDate).isSameOrAfter(moment(), 'day');
      case 'past':
        return moment(trip.endDate).isBefore(moment(), 'day');
      case 'all':
        return true;
    }
  }).sort((a, b) => {
    return a.startDate < b.startDate ? 1 : -1;
  });
};