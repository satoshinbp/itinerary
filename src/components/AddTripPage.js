import React from 'react';
import { connect } from 'react-redux';
import TripForm from './TripForm';
import { addTrip } from '../actions/trips';

const AddTripPage = (props) => (
  <div className="container">
    <TripForm
      onSubmit={(trip) => {
        props.dispatch(addTrip(trip));
        props.history.push('/');
      }} />
  </div>
);

export default connect()(AddTripPage);