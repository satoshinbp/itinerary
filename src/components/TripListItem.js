import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { removeTrip } from '../actions/trips'

const TripListItem = ({ dispatch, id, title, startDate, endDate, setTripId }) => {
  const confirmRemoveTrip = () => {
    const result = confirm('Are you sure you want to delete this trip?');

    if (result) {
      dispatch(removeTrip({ id }))
    }
  }

  return (
    <div className="trip-list-item">
      <div className="trip-list-item__actions">
        <FontAwesomeIcon
          className="icon"
          icon="edit"
          onClick={() => setTripId(id)}
        />
        <FontAwesomeIcon
          className="icon"
          icon="trash-alt"
          onClick={confirmRemoveTrip}
        />
      </div>
      <Link className="trip-list-item__title" to={`/schedule/${id}`}>{title}</Link>
      <Link className="trip-list-item__dates" to={`/schedule/${id}`}>
        {
          startDate.isSame(endDate, 'day') ? (
            <span>
              {`${startDate.format('YYYY/MM/DD')}`}
            </span>
          ) : (
              <span>
                {`${startDate.format('YYYY/MM/DD')} - ${endDate.format('YYYY/MM/DD')}`}
              </span>
            )
        }
      </Link>
    </div>
  )
}

export default connect()(TripListItem)