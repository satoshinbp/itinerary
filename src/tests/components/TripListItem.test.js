import React from 'react'
import { shallow } from 'enzyme'
import { TripListItem } from '../../components/TripListItem'
import { removeTrip } from '../../actions/trips'
import trips from '../fixtures/trips'

let dispatch, setTripId, wrapper

beforeEach(() => {
  setTripId = jest.fn()
  dispatch = jest.fn()
})

test('should render TripListItem for multi days trip', () => {
  wrapper = shallow(<TripListItem dispatch={dispatch} {...trips[0]} setTripId={setTripId} />)
  expect(wrapper).toMatchSnapshot()
})

test('should render TripListItem for one day trip', () => {
  wrapper = shallow(<TripListItem dispatch={dispatch} {...trips[2]} setTripId={setTripId} />)
  expect(wrapper).toMatchSnapshot()
})

test('should set tripId when edit icon clicked', () => {
  wrapper = shallow(<TripListItem dispatch={dispatch} {...trips[0]} setTripId={setTripId} />)
  wrapper.find({ icon: 'edit' }).simulate('click')
  expect(setTripId).toBeCalledWith(trips[0].id)
})

test('should remove trip when trash icon clicked', () => {
  wrapper = shallow(<TripListItem dispatch={dispatch} {...trips[0]} setTripId={setTripId} />)
  window.confirm = jest.fn(() => true)
  wrapper.find({ icon: 'trash-alt' }).simulate('click')
  expect(dispatch).toBeCalledWith(removeTrip(trips[0].id))
})