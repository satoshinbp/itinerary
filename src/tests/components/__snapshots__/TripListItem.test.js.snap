// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TripListItem for multi days trip 1`] = `
<div
  className="trip-list-item"
>
  <div
    className="trip-list-item__actions"
  >
    <FontAwesomeIcon
      border={false}
      className="icon"
      fixedWidth={false}
      flip={null}
      icon="edit"
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    <FontAwesomeIcon
      border={false}
      className="icon"
      fixedWidth={false}
      flip={null}
      icon="trash-alt"
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </div>
  <Link
    className="trip-list-item__title"
    to="/schedule/a"
  >
    Summer Vacation in Hawai
  </Link>
  <Link
    className="trip-list-item__dates"
    to="/schedule/a"
  >
    <span>
      1970/01/01 - 1970/01/04
    </span>
  </Link>
</div>
`;

exports[`should render TripListItem for one day trip 1`] = `
<div
  className="trip-list-item"
>
  <div
    className="trip-list-item__actions"
  >
    <FontAwesomeIcon
      border={false}
      className="icon"
      fixedWidth={false}
      flip={null}
      icon="edit"
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    <FontAwesomeIcon
      border={false}
      className="icon"
      fixedWidth={false}
      flip={null}
      icon="trash-alt"
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </div>
  <Link
    className="trip-list-item__title"
    to="/schedule/c"
  >
    Tokyo - 2020 Spring
  </Link>
  <Link
    className="trip-list-item__dates"
    to="/schedule/c"
  >
    <span>
      1969/12/29
    </span>
  </Link>
</div>
`;
